packages/github/README.md

```diff
@@ -17,7 +17,13 @@ async function run() {
     // https://help.github.com/en/actions/automating-your-workflow-with-github-actions/authenticating-with-the-github_token#about-the-github_token-secret
     const myToken = core.getInput('myToken');
 
-    const octokit = new github.GitHub(myToken);
+    const octokit = new github.GitHub(github.getOptions(myToken));
```

Comment:
have you considered an interface like:

```javascript
export function newOctokit(token: string, options?: OctokitOptions, plugin: SOME_TYPE[]): Octokit
```

Advantages:
- Simplify construction
  - `const octokit = github.newOctokit(myToken)`<br />instead of<br />`const octokit = new github.GitHub(github.getOptions(myToken))`
- Simplify by eliminating inheritance (i dont think we add or override any functions?)
- Support plugins?
  - Honestly dont understand how the plugin stuff works with typings. If allowed optional `plugins` arg (assume could use [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) to pass to the plugins function), would the consumer need to cast to whatever type they expect?

Reply:
Also would enable customers to more easily mock

Reply:
Sorry just noticed the comments on the conversation tab... ok didnt consider `.plugin` example you had:

```
const myGitHub = GitHub.plugin(myPlugin)
[...]
const octokit = new myGitHub(getOptions(token), {//whatever octokit options you want})
```

In that case, the defaults wouldnt be set though, correct? Late right now. Let's touch base in the morning.

Reply:
The defaults will get set in the plugin scenario. You could also override the default and/or add additional defaults
```
// Always use the baseurl for dotcom, set a default timezone as well.
const myGitHub = GitHub.plugin(myplugin).defaults({
  baseUrl: "https://github.com/api",
  timeZone: "America/Los_Angeles"
})

const myInstance = new myGitHub(getOptions(token))
```

getInstance has a hard time returning the type, because the type is generated when you do `octokit.plugin` or `octokit.defaults`

So a function like

```
export function getInstance(
  token: string,
  options?: OctokitOptions,
  plugins?: OctokitPlugin[]
): Octokit {
  var myClass;
  if (plugins != null){
    myClass = Octokit.plugin(restEndpointMethods, paginateRest, ...plugins).defaults(defaults)
  }
  else {
    myClass = Octokit.plugin(restEndpointMethods, paginateRest).defaults(defaults)
  }
  return new myClass(getOptions(token, options))
}
```

Doesn't have great type information attached. For example, you can run 
```
const myOctokit = getInstance(token)
myOctokit.repos.getBranch({
        owner: 'actions',
        repo: 'toolkit',
        branch: 'master'
})
```

But you we don't get the type information to support it, so consumers will just have to know what functions exist which isn't a great UX.

There might be a way to return typeOf GitHub, but the returned object still would not have the type information associated with the plugins they pass in, so I don't really think we should go down this route.

Reply:
Consider optimizing for simple case (assumption this is the mainline):

```javascript
// simple case
export function createOctokit(token: string, options?: OctokitOptions)
```

And guidance for advanced users:

```javascript
const octokitType = Octokit.plugin([...])
  .defaults(github.createDefaults())
const octokit = new octokitType(github.createOptions(token, opts))
```

Reply:
Done!